client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -144,13 +155,39 @@
         List<ByteBuffer> data = events.stream().map(serializer::serialize).collect(Collectors.toList());
         CompletableFuture<Void> ackFuture = new CompletableFuture<Void>();
         synchronized (writeFlushLock) {
-            synchronized (writeSealLock) {
-                SegmentOutputStream segmentWriter = getSegmentWriter(routingKey);
-                segmentWriter.write(PendingEvent.withHeader(routingKey, data, ackFuture));
+            if (data.stream().mapToInt(m -> m.remaining()).sum() > Serializer.MAX_EVENT_SIZE) {
+                writeLargeEvent(routingKey, data, ackFuture);
+            } else {
+                synchronized (writeSealLock) {
+                    SegmentOutputStream segmentWriter = getSegmentWriter(routingKey);
+                    segmentWriter.write(PendingEvent.withHeader(routingKey, data, ackFuture));
+                }
             }
         }
         return ackFuture;
     }
+    
+    @GuardedBy("writeFlushLock")
+    private void writeLargeEvent(String routingKey, List<ByteBuffer> events, CompletableFuture<Void> ackFuture) {
+        flush();
+        boolean success = false;
+        LargeEventWriter writer = new LargeEventWriter(UUID.randomUUID(), controller, connectionPool);
+        while (!success) {
+            Segment segment = selector.getSegmentForEvent(routingKey);
+            try {
+                writer.writeLargeEvent(segment, events, tokenProvider, config);
+                success = true;
+                ackFuture.complete(null);
+            } catch (SegmentSealedException | NoSuchSegmentException e) {
+                log.warn("Write large event on segment {} failed due to {}, it will be retried.", segment, e.getMessage());
+                handleLogSealed(segment);
```

Comment:
`getUnackedEventsOnSeal ` would not return these events in case of a sealed segment. Is there a different way by which these events are retransmitted?

Reply:
The partial large event data won't be re-written as part of handle log sealed. However this try block is surrounded by a while loop which will re-write the data.